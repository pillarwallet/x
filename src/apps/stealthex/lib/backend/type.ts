export type UnavailableResponse = {
    state: 'unknown';
  };
  
export type FailedResponse<T> = {
    state: 'failed';
    status: number;
    message?: T;
};

export type ThrowableError<T> = UnavailableResponse | FailedResponse<T>;

export type ErrorResponse<T> = ThrowableError<T> & { data: null };

export type ThrowableResponse<T> = {
    status: number;
    data: T;
    headers: Headers;
};

export type SuccessResponse<T> = ThrowableResponse<T> & {
state: 'success';
};
  
export type Response<D, E> = SuccessResponse<D> | ErrorResponse<E>;

export type RequestResponse<
  T extends boolean,
  D = unknown,
  E = string,
> = T extends true ? ThrowableResponse<D> : Response<D, E>;