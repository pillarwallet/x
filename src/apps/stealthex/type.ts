import { CurrencyData } from './lib/backend/api';

export type Config = {
    amount: string;
    fixed: boolean;
    fiat: boolean;
    sendCurrency: CurrencyData;
    receiveCurrency: CurrencyData;
    getPairs?: CurrencyData[];
    sendPairs?: CurrencyData[];
    recepientAddress?: string;
    lockReceiveCurrency?: boolean;
}

export type ExchangeInfoCurrencyPair = {
    [key: string]: CurrencyData | undefined;
  };

export type ExchangeInfo = {
    id: string;
    type: string;
    timestamp: string;
    updated_at: string;
    currency_from: string;
    currency_to: string;
    amount_from: string;
    expected_amount: string;
    amount_to: string;
    amount_to_btc: string | null;
    address_from: string;
    address_to: string;
    extra_id_from: string | null;
    extra_id_to: string | null;
    tx_from: string | null;
    tx_to: string | null;
    status:
      | 'waiting'
      | 'confirming'
      | 'exchanging'
      | 'sending'
      | 'finished'
      | 'failed'
      | 'refunded'
      | 'expired'
      | 'verifying';
    refund_address: string | null;
    refund_extra_id: string | null;
    currencies: ExchangeInfoCurrencyPair;
    review?: boolean;
    payment_id?: string;
    mercuryo_signature: string;
};

export type StructuredCurrency = {
  fiat: CurrencyData[];
  popular: CurrencyData[];
  other: CurrencyData[];
};