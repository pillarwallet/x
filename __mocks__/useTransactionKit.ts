import { vi } from 'vitest';

const useTransactionKit = vi.fn(() => ({
  kit: {
    getState: vi.fn(() => ({
      namedTransactions: {},
      batches: {},
      isEstimating: false,
      isSending: false,
      containsSendingError: false,
      containsEstimatingError: false,
    })),
    getEtherspotProvider: vi.fn(() => ({
      getChainId: vi.fn(() => 1),
    })),
    transaction: vi.fn(() => ({
      name: vi.fn(() => ({
        estimate: vi.fn(() => Promise.resolve({})),
        send: vi.fn(() => Promise.resolve({})),
      })),
    })),
    estimateBatches: vi.fn(() => Promise.resolve({})),
    sendBatches: vi.fn(() => Promise.resolve({})),
  },
  walletAddress: '0x1234567890123456789012345678901234567890',
  setWalletAddress: vi.fn(),
}));

export default useTransactionKit;
