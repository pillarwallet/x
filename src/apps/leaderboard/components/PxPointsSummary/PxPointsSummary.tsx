import { useWalletAddress } from '@etherspot/transaction-kit';

// types
import { LeaderboardTableData } from '../../../../types/api';

// images
import LotteryTicketIcon from '../../images/lottery-qualified-icon.png';
import MigrationIcon from '../../images/migration-icon.svg';
import PillarWalletIcon from '../../images/pillar-wallet-icon.svg';
import TradingIcon from '../../images/trading-icon.svg';

// components
import GasNewDropCard from '../PointsCards/GasNewDropCard';
import OverviewPointsCard from '../PointsCards/OverviewPointsCard';
import PointsCard from '../PointsCards/PointsCard';
import BodySmall from '../Typography/BodySmall';

type PxPointsSummaryProps = {
  allTimeTradingData: LeaderboardTableData[] | undefined;
  allTimeMigrationData: LeaderboardTableData[] | undefined;
  mergedAllTimeData: LeaderboardTableData[] | undefined;
  mergedWeeklyTimeData: LeaderboardTableData[] | undefined;
  isUserInMigrationData: boolean;
};

const PxPointsSummary = ({
  allTimeTradingData,
  allTimeMigrationData,
  mergedAllTimeData,
  mergedWeeklyTimeData,
  isUserInMigrationData,
}: PxPointsSummaryProps) => {
  const walletAddress = useWalletAddress();

  const findMatchingEntry = (data?: LeaderboardTableData[]) => {
    if (!walletAddress || !data) return { entry: undefined, index: -1 };

    const index = data.findIndex((entry) =>
      entry.addresses.some(
        (addr) => addr.toLowerCase() === walletAddress.toLowerCase()
      )
    );

    return {
      entry: index !== -1 ? data[index] : undefined,
      index,
    };
  };

  const myAllTimeTrading = findMatchingEntry(allTimeTradingData);
  const myAllTimeMigration = findMatchingEntry(allTimeMigrationData);
  const myAllTimeMerged = findMatchingEntry(mergedAllTimeData);
  const myWeeklyMerged = findMatchingEntry(mergedWeeklyTimeData);

  const isEligibleForLottery = myWeeklyMerged.entry?.completedSwap === true;

  return (
    <div className="flex flex-col bg-container_grey desktop:py-5 desktop:px-6 tablet:py-5 tablet:px-6 mobile:p-3 rounded-2xl">
      <div className="flex justify-between items-center">
        <div className="flex gap-2 items-center">
          {isUserInMigrationData && (
            <div className="w-5 h-5 rounded p-[0.6px] bg-gradient-to-b from-white via-white/[.6] to-[#360064] shadow-[0_0_15px_0_#693397,0_2px_6px_0_#00000080]">
              <div
                className="flex w-full h-full items-center justify-center rounded bg-[#360064]"
                style={{
                  backgroundImage:
                    'radial-gradient(100% 100% at 50% 0%, rgba(255, 255, 255, 0.7) 0%, rgba(255, 255, 255, 0) 100%)',
                }}
              >
                <img src={PillarWalletIcon} alt="pillar-wallet-icon" />
              </div>
            </div>
          )}

          <p className="desktop:text-xl tablet:text-xl mobile:text-base font-medium">
            My PX Points
          </p>
        </div>
        {isEligibleForLottery && (
          <div className="flex gap-1 items-center px-1.5 py-[4px] rounded-md bg-[#25232D]">
            <img
              src={LotteryTicketIcon}
              alt="lottery-qualified-icon"
              className="w-5 h-4"
            />
            <BodySmall className="text-percentage_green font-normal">
              Qualified{' '}
              <span className="desktop:inline tablet:inline mobile:hidden">
                for Lottery
              </span>
            </BodySmall>
          </div>
        )}
      </div>

      <div className="desktop:flex tablet:flex mobile:flex mobile:flex-col gap-3 mt-4">
        <div className="desktop:w-[44%] tablet:w-[44%] mobile:w-full">
          <OverviewPointsCard
            myAllTimeMerged={myAllTimeMerged}
            myWeeklyMerged={myWeeklyMerged}
          />
        </div>

        <div
          className={`desktop:w-[56%] tablet:w-[56%] mobile:w-full grid ${isUserInMigrationData ? 'grid-cols-2' : 'grid-cols-1'} gap-3`}
        >
          {isUserInMigrationData && myAllTimeMigration && (
            <PointsCard
              title="Migration"
              icon={MigrationIcon}
              background="bg-percentage_red/[.05]"
              points={myAllTimeMigration.entry?.totalPoints || 0}
              rank={
                myAllTimeMigration.index !== -1
                  ? myAllTimeMigration.index + 1
                  : undefined
              }
              textTooltip="These are points calculated from the amount that you have transferred from your Pillar V1 and Pillar V2 wallets to PX."
            />
          )}
          <PointsCard
            title="Trading"
            icon={TradingIcon}
            background="bg-percentage_green/[.05]"
            points={myAllTimeTrading.entry?.totalPoints || 0}
            rank={
              myAllTimeTrading.index !== -1
                ? myAllTimeTrading.index + 1
                : undefined
            }
            textTooltip="These are points calculated from trading assets, only trades above $2 are calculated."
          />
          <GasNewDropCard
            newDropTime={0}
            gasUsed={myAllTimeMigration.entry?.totalGas || 0}
            isMigrated
          />
        </div>
      </div>
    </div>
  );
};

export default PxPointsSummary;
