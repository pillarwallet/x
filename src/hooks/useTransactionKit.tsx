import { useContext } from 'react';

// providers
import { EtherspotTransactionKitContext } from '../providers/EtherspotTransactionKitProvider';

const useTransactionKit = () => {
  const context = useContext(EtherspotTransactionKitContext);

  if (context === null) {
    throw new Error('No parent <EtherspotTransactionKitProvider />');
  }

  return context?.data;
};

export default useTransactionKit;
