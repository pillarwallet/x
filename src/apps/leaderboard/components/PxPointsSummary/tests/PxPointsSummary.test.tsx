import { render, screen } from '@testing-library/react';
import { vi } from 'vitest';

// types
import type { Mock } from 'vitest';
import {
  LeaderboardRankChange,
  LeaderboardTableData,
} from '../../../../../types/api';

// components
import PxPointsSummary from '../PxPointsSummary';
import useTransactionKit from '../../../../../hooks/useTransactionKit';

// mock subcomponents
vi.mock('../../PointsCards/GasNewDropCard', () => ({
  __esModule: true,
  default: function GasNewDropCard() {
    return <div>GasNewDropCard</div>;
  },
}));

vi.mock('../../PointsCards/OverviewPointsCard', () => ({
  __esModule: true,
  default: function OverviewPointsCard() {
    return <div>OverviewPointsCard</div>;
  },
}));

vi.mock('../../PointsCards/PointsCard', () => ({
  __esModule: true,
  default: function PointsCard(props: { title: string }) {
    return <div>{props.title} PointsCard</div>;
  },
}));

vi.mock('../../Typography/BodySmall', () => ({
  __esModule: true,
  default: function BodySmall(props: { children: React.ReactNode }) {
    return <div>{props.children}</div>;
  },
}));

vi.mock('../../../../../hooks/useTransactionKit');

describe('<PxPointsSummary />', () => {
  const mockWallet = '0x1234';
  const useTransactionKitMock = useTransactionKit as unknown as Mock;

  const allTimeTradingDataMock: LeaderboardTableData[] = [
    {
      totalPoints: 1200,
      totalAmountUsd: 500000,
      addresses: ['0x1234', '0x456def'],
      totalGas: 35000,
    },
    {
      totalPoints: 1100,
      totalAmountUsd: 450000,
      addresses: ['0x789ghi'],
      totalGas: 30000,
    },
  ];

  const allTimeMigrationDataMock: LeaderboardTableData[] = [
    {
      totalPoints: 900,
      totalAmountUsd: 200000,
      addresses: ['0x1234'],
      totalGas: 15000,
    },
    {
      totalPoints: 850,
      totalAmountUsd: 180000,
      addresses: ['0xdef456', '0xghi789'],
      totalGas: 14000,
    },
  ];

  const mergedAllTimeDataMock: LeaderboardTableData[] = [
    {
      totalPoints: 2100,
      totalAmountUsd: 700000,
      addresses: ['0x1234', '0x789ghi'],
      totalGas: 50000,
    },
    {
      totalPoints: 1950,
      totalAmountUsd: 630000,
      addresses: ['0xdef456', '0x456def'],
      totalGas: 44000,
    },
  ];

  const mergedWeeklyTimeDataMock: LeaderboardTableData[] = [
    {
      totalPoints: 350,
      totalAmountUsd: 120000,
      addresses: ['0x1234'],
      completedSwap: true,
      totalGas: 8000,
      rankChange: LeaderboardRankChange.INCREASED,
    },
    {
      totalPoints: 300,
      totalAmountUsd: 100000,
      addresses: ['0xdef456'],
      completedSwap: false,
      totalGas: 7500,
      rankChange: LeaderboardRankChange.NO_CHANGE,
    },
  ];

  beforeEach(() => {
    useTransactionKitMock.mockReturnValue({
      walletAddress: mockWallet,
      kit: {},
      activeChainId: 1,
      setActiveChainId: vi.fn(),
    });
  });

  it('renders correctly and matches snapshot', () => {
    const { container } = render(
      <PxPointsSummary
        allTimeTradingData={allTimeTradingDataMock}
        allTimeMigrationData={allTimeMigrationDataMock}
        mergedAllTimeData={mergedAllTimeDataMock}
        mergedWeeklyTimeData={mergedWeeklyTimeDataMock}
        isUserInMigrationData
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('renders full Px Points summary with all sections when user is present in all data sets', () => {
    render(
      <PxPointsSummary
        allTimeTradingData={allTimeTradingDataMock}
        allTimeMigrationData={allTimeMigrationDataMock}
        mergedAllTimeData={mergedAllTimeDataMock}
        mergedWeeklyTimeData={mergedWeeklyTimeDataMock}
        isUserInMigrationData
      />
    );

    expect(screen.getByText('My PX Points')).toBeInTheDocument();
    expect(screen.getByText('Qualified')).toBeInTheDocument();
    expect(screen.getByText('OverviewPointsCard')).toBeInTheDocument();
    expect(screen.getByText('Migration PointsCard')).toBeInTheDocument();
    expect(screen.getByText('Trading PointsCard')).toBeInTheDocument();
    expect(screen.getByText('GasNewDropCard')).toBeInTheDocument();
  });

  it('does not show migration card if isUserInMigrationData is false', () => {
    render(
      <PxPointsSummary
        allTimeTradingData={allTimeTradingDataMock}
        allTimeMigrationData={[]}
        mergedAllTimeData={mergedAllTimeDataMock}
        mergedWeeklyTimeData={mergedWeeklyTimeDataMock}
        isUserInMigrationData={false}
      />
    );

    expect(screen.getByText('My PX Points')).toBeInTheDocument();
    expect(screen.queryByText('Migration PointsCard')).not.toBeInTheDocument();
    expect(screen.getByText('Trading PointsCard')).toBeInTheDocument();
  });

  it('does not show lottery badge if user is not eligible', () => {
    const newMergedWeeklyTimeDataMock = mergedWeeklyTimeDataMock.map(
      (entry, index) =>
        index === 0 ? { ...entry, completedSwap: false } : entry
    );

    render(
      <PxPointsSummary
        allTimeTradingData={allTimeTradingDataMock}
        allTimeMigrationData={[]}
        mergedAllTimeData={mergedAllTimeDataMock}
        mergedWeeklyTimeData={newMergedWeeklyTimeDataMock}
        isUserInMigrationData={false}
      />
    );

    expect(screen.queryByText('Qualified')).not.toBeInTheDocument();
  });

  it('handles no wallet address gracefully', () => {
    useTransactionKitMock.mockReturnValue({
      walletAddress: undefined,
      kit: {},
      activeChainId: 1,
      setActiveChainId: vi.fn(),
    });

    render(
      <PxPointsSummary
        allTimeTradingData={[]}
        allTimeMigrationData={[]}
        mergedAllTimeData={[]}
        mergedWeeklyTimeData={[]}
        isUserInMigrationData={false}
      />
    );

    expect(screen.getByText('My PX Points')).toBeInTheDocument();
    expect(screen.getByText('OverviewPointsCard')).toBeInTheDocument(); // still renders overview
    expect(screen.getByText('Trading PointsCard')).toBeInTheDocument(); // with 0 points
  });
});
