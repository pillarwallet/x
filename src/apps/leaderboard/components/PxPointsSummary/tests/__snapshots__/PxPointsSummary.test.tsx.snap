// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PxPointsSummary /> > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col bg-container_grey desktop:py-5 desktop:px-6 tablet:py-5 tablet:px-6 mobile:p-3 rounded-2xl"
  >
    <div
      class="flex justify-between items-center"
    >
      <div
        class="flex gap-2 items-center"
      >
        <div
          class="w-5 h-5 rounded p-[0.6px] bg-gradient-to-b from-white via-white/[.6] to-[#360064] shadow-[0_0_15px_0_#693397,0_2px_6px_0_#00000080]"
        >
          <div
            class="flex w-full h-full items-center justify-center rounded bg-[#360064]"
          >
            <img
              alt="pillar-wallet-icon"
              src="data:image/svg+xml,%3csvg%20width='10'%20height='14'%20viewBox='0%200%2010%2014'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20d='M6.86766%204.75968V8.62798H3.76708H0.666504V9.78571V10.9409L1.82423%2012.0987C2.45936%2012.7338%202.98698%2013.2539%202.99945%2013.2539C3.00948%2013.2539%203.01693%2012.7363%203.01693%2012.1037V10.9535H6.10504H9.19316V7.08275V3.21437L8.03041%202.05162L6.86766%200.888875V4.75968Z'%20fill='white'/%3e%3cpath%20d='M0.666504%203.97711V5.52744H2.2293H3.7921V3.97711V2.42678H2.2293H0.666504V3.97711Z'%20fill='white'/%3e%3c/svg%3e"
            />
          </div>
        </div>
        <p
          class="desktop:text-xl tablet:text-xl mobile:text-base font-medium"
        >
          My PX Points
        </p>
      </div>
      <div
        class="flex gap-1 items-center px-1.5 py-[4px] rounded-md bg-[#25232D]"
      >
        <img
          alt="lottery-qualified-icon"
          class="w-5 h-4"
          src="/src/apps/leaderboard/images/lottery-qualified-icon.png"
        />
        <div>
          Qualified
           
          <span
            class="desktop:inline tablet:inline mobile:hidden"
          >
            for Lottery
          </span>
        </div>
      </div>
    </div>
    <div
      class="desktop:flex tablet:flex mobile:flex mobile:flex-col gap-3 mt-4"
    >
      <div
        class="desktop:w-[44%] tablet:w-[44%] mobile:w-full"
      >
        <div>
          OverviewPointsCard
        </div>
      </div>
      <div
        class="desktop:w-[56%] tablet:w-[56%] mobile:w-full grid grid-cols-2 gap-3"
      >
        <div>
          Migration
           PointsCard
        </div>
        <div>
          Trading
           PointsCard
        </div>
        <div>
          GasNewDropCard
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PxPointsSummary /> renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col bg-container_grey desktop:py-5 desktop:px-6 tablet:py-5 tablet:px-6 mobile:p-3 rounded-2xl"
  >
    <div
      class="flex justify-between items-center"
    >
      <div
        class="flex gap-2 items-center"
      >
        <div
          class="w-5 h-5 rounded p-[0.6px] bg-gradient-to-b from-white via-white/[.6] to-[#360064] shadow-[0_0_15px_0_#693397,0_2px_6px_0_#00000080]"
        >
          <div
            class="flex w-full h-full items-center justify-center rounded bg-[#360064]"
          >
            <img
              alt="pillar-wallet-icon"
              src="[object Object]"
            />
          </div>
        </div>
        <p
          class="desktop:text-xl tablet:text-xl mobile:text-base font-medium"
        >
          My PX Points
        </p>
      </div>
      <div
        class="flex gap-1 items-center px-1.5 py-[4px] rounded-md bg-[#25232D]"
      >
        <img
          alt="lottery-qualified-icon"
          class="w-5 h-4"
          src="lottery-qualified-icon.png"
        />
        <div>
          Qualified
           
          <span
            class="desktop:inline tablet:inline mobile:hidden"
          >
            for Lottery
          </span>
        </div>
      </div>
    </div>
    <div
      class="desktop:flex tablet:flex mobile:flex mobile:flex-col gap-3 mt-4"
    >
      <div
        class="desktop:w-[44%] tablet:w-[44%] mobile:w-full"
      >
        <div>
          OverviewPointsCard
        </div>
      </div>
      <div
        class="desktop:w-[56%] tablet:w-[56%] mobile:w-full grid grid-cols-2 gap-3"
      >
        <div>
          Migration
           PointsCard
        </div>
        <div>
          Trading
           PointsCard
        </div>
        <div>
          GasNewDropCard
        </div>
      </div>
    </div>
  </div>
</div>
`;
